File: frontend/src/types/businessOperationMapping.ts
/**
 * Business Operation Mapping Types
 * Типы для маппинга хозяйственных операций из 1С на категории бюджета
 */

export interface BusinessOperationMapping {
  id: number
  business_operation: string
  category_id: number
  category_name?: string
  category_type?: 'OPEX' | 'CAPEX'
  priority: number
  confidence: number
  notes?: string
  department_id: number
  department_name?: string
  is_active: boolean
  created_at: string
  updated_at?: string
}

export interface BusinessOperationMappingList {
  items: BusinessOperationMapping[]
  total: number
  skip: number
  limit: number
}

export interface BusinessOperationMappingCreate {
  business_operation: string
  category_id: number
  priority?: number
  confidence?: number
  notes?: string
  department_id: number
}

export interface BusinessOperationMappingUpdate {
  category_id?: number
  priority?: number
  confidence?: number
  notes?: string
  is_active?: boolean
}

export interface BusinessOperationMappingFilters {
  department_id?: number
  is_active?: boolean
  business_operation?: string
  skip?: number
  limit?: number
}

File: backend/app/schemas/business_operation_mapping.py
"""
Pydantic schemas for Business Operation Mappings
"""
from typing import Optional, List
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, Field, field_validator


class BusinessOperationMappingBase(BaseModel):
    """Base schema for business operation mapping"""
    business_operation: str = Field(..., description="Business operation name from 1C", max_length=100)
    category_id: Optional[int] = Field(None, description="Budget category ID (nullable for auto-created stubs)")
    priority: int = Field(default=10, ge=1, le=100, description="Priority (higher = more important)")
    confidence: Decimal = Field(default=Decimal("0.98"), ge=0, le=1, description="Confidence level (0-1)")
    notes: Optional[str] = Field(None, description="Optional notes")

    @field_validator('confidence', mode='before')
    @classmethod
    def validate_confidence(cls, v):
        if isinstance(v, str):
            return Decimal(v)
        if isinstance(v, (int, float)):
            return Decimal(str(v))
        return v


class BusinessOperationMappingCreate(BusinessOperationMappingBase):
    """Schema for creating business operation mapping"""
    department_id: int = Field(..., description="Department ID")


class BusinessOperationMappingUpdate(BaseModel):
    """Schema for updating business operation mapping"""
    category_id: Optional[int] = None
    priority: Optional[int] = Field(None, ge=1, le=100)
    confidence: Optional[Decimal] = Field(None, ge=0, le=1)
    notes: Optional[str] = None
    is_active: Optional[bool] = None

    @field_validator('confidence', mode='before')
    @classmethod
    def validate_confidence(cls, v):
        if v is None:
            return v
        if isinstance(v, str):
            return Decimal(v)
        if isinstance(v, (int, float)):
            return Decimal(str(v))
        return v


class BusinessOperationMappingInDB(BaseModel):
    """Schema for business operation mapping in database"""
    id: int
    business_operation: str
    category_id: Optional[int] = None
    category_name: Optional[str] = None
    category_type: Optional[str] = None
    priority: int
    confidence: float
    notes: Optional[str] = None
    department_id: int
    department_name: Optional[str] = None
    is_active: bool
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class BusinessOperationMappingList(BaseModel):
    """Schema for paginated list of mappings"""
    items: List[BusinessOperationMappingInDB]
    total: int
    skip: int
    limit: int
